topdown/eval.go

```diff
@@ -2502,6 +2521,55 @@ func (e evalVirtualPartial) evalOneRulePreUnify(iter unifyIterator, rule *ast.Ru
 	return nil
 }
 
+func (e evalVirtualPartial) evalOneDynamicRefRulePreUnify(rule *ast.Rule, result *ast.Term, unknown bool) (*ast.Term, error) {
+
+	child := e.e.child(rule.Body)
+
+	child.traceEnter(rule)
+	var defined bool
+
+	// Walk the dynamic portion of rule ref to unify vars
+	err := child.biunifyDynamicRef(e.pos+1, rule.Ref(), e.ref, child.bindings, e.bindings, func() error {
+		defined = true
+		return child.eval(func(child *eval) error {
+			child.traceExit(rule)
+			var dup bool
+			var err error
+			result, dup, err = e.reduce(rule, child.bindings, result)
+			if err != nil {
+				return err
+			} else if !unknown && dup {
+				child.traceDuplicate(rule)
+				return nil
+			}
+
+			child.traceRedo(rule)
+			return nil
+		})
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	// We're tracing here to exhibit similar behaviour to evalOneRulePreUnify
+	if !defined {
+		child.traceFail(rule)
+	}
+
+	return result, nil
+}
+
+func (e *eval) biunifyDynamicRef(pos int, a, b ast.Ref, b1, b2 *bindings, iter unifyIterator) error {
+	if pos >= len(a) || pos >= len(b) {
+		return iter()
```

Comment:
So if they don't have the same length, but overlap, we'd accept that?
```rego
data.foo.bar = data.foo[x].z
```
we'd add `x = "bar"` to the bindings and _go on_?

But I might be missing something here, I haven't fully understood the topdown change yet.

Reply:
We're simply binding any possibly known vars in a rule's ref before evaluating its body.

If `data.foo.bar == data.foo.bar.z` then I'd expect your snipped should evaluate. (is that even possible, though? 😵‍💫)

This is an interesting case to test for, though; hadn't thought about it from this angle before 🤔.

Perhaps a bit hacky test, but this should trigger this scenario:

```rego
package unification

p := i {
    foo.bar.x.y = foo[i].z
}

foo[a][b] := c {
    a := "bar"
    c := {
        "x": {
            "y": 42
        },
        "z": 42
    }[b]
}
```
```bash
$ OPA_ENABLE_GENERAL_RULE_REFS=true opa eval -d . 'data.unification.p'
```

Reply:
Or am I completely misunderstanding you?